### DISCUSSION TITLE
watcher on async actions not automatically stopped on vue 3.5 (while vue 3.4 works fine)

### DISCUSSION THREAD

**Discussion by natuan62** [vuejs/pinia#2787] [Help and Questions]
### Reproduction

source in `vue-3.5-pinia-test.zip` (sorry i try but still can't config source on a reproduction link)

[vue-3.5-pinia-test.zip](https://github.com/user-attachments/files/17239250/vue-3.5-pinia-test.zip)

### Steps to reproduce the bug

I not sure this is vue 3.5 bug or pinia (because vue 3.4 still work fine)

- npm install
- npm run dev

```javascript
store.ts

import { defineStore } from 'pinia';
import { computed, watch } from 'vue';

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export const useDesignStore = defineStore('designEditorStore', {
  state: () => ({
    designId: 0,
  }),
  actions: {
    async init() {
      const design = computed(() => this.router.currentRoute.value.params.designId as string);

      await sleep(1000);

      watch(design, () => {
        console.log('14 watch', design.value);
        this.designId = Number(design.value);
      });
    },
  },
});

```

```javascript
// router file

import { createRouter, createWebHistory } from 'vue-router';
import HelloWorld from '../components/HelloWorld.vue';
import Home from '../components/Home.vue';

const routes = [
  { path: '/design/:designId', component: HelloWorld, name: 'Design' },
  { path: '/home', component: Home, name: 'Home' },
  { path: '/', component: Home, name: 'Home' },
];

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [...routes],
});

export default router;

```

```javascript
App.vue

<template>
  <button @click="goDesign()">go design</button>
  <router-view />
</template>

<script>
import { useDesignStore } from './store';
import HelloWorld from './components/HelloWorld.vue';
import { useRouter } from 'vue-router';
import router from './router';
export default {
  components: {
    HelloWorld,
  },
  setup() {
    const router = useRouter();

    const goDesign = (num) => {
      router.push({ name: 'Design', params: { designId: 0 } });
    };

    return {
      goDesign,
    };
  },
};
</script>

```

```javascript
HelloWorld.vue

<template>
  <button @click="goHome()" style="margin: 5px">out</button>
  <p>HelloWorld</p>
  <div v-for="item in designIds" :key="item" style="margin: 5px">
    <button @click="viewDesign(item)">Design {{ item }}</button>
  </div>
</template>

<script>
import { reactive, onMounted, toRefs, onUnmounted, watch, computed } from 'vue';
import { useDesignStore } from '../store';
import { useRoute, useRouter } from 'vue-router';

export default {
  setup() {
    const router = useRouter();
    const designIds = [1, 2];
    const useStore = useDesignStore();

    const design = computed(() => router.currentRoute.value.params.designId);

    const viewDesign = (id) => {
      router.push({ name: 'Design', params: { designId: id } });
    };

    const goHome = () => {
      router.push({ name: 'Home' });
    };

    onMounted(async () => {
      console.log('35 onMounted HelloWorld');
      await useStore.init();
    });
    onUnmounted(() => {
      console.log('40 onUnmounted HelloWorld');
    });

    return {
      designIds,
      viewDesign,
      goHome,
    };
  },
};
</script>

```

[vue-3.5-pinia-test.zip](https://github.com/user-attachments/files/17239250/vue-3.5-pinia-test.zip)

### Expected behavior

- watch not trigger many times like this demo

### Actual behavior

- watch trigger many times when mount component

### Additional information

https://github.com/user-attachments/assets/49ded383-d948-49ec-90ce-e041e4e6fcc5

**Comment by natuan62** [comment]
Hi @posva if i moved watch to up, it works fine

```javascript
async init() {
      const design = computed(() => this.router.currentRoute.value.params.designId as string);

      watch(design, () => {
        console.log('14 watch', design.value);
        this.designId = Number(design.value);
      });

     await sleep(1000);
    },
```

**Comment by ORCHORDS** [comment]
Confirmed root cause. This is the exact behavior change between Vue 3.4 and 3.5. The Vue 3.5 changelog has the relevant fixes:

> "ssr: prevent watch from firing after async setup await (#14547)" — 3.5.30
> "reactivity: ensure multiple effectScope on() and off() calls maintains correct active scope" — 3.5.14
> "reactivity: release nested effects/scopes on effect scope stop (#12373)" — 3.5.13
> "watch: watchEffect clean-up with SSR (#12097)" — 3.5.12

What actually changed and why your watcher fires many times:

1. Pinia actions wrap their execution in an effect scope. Any reactive effects (refs, computeds, watch callbacks) created synchronously inside an action are auto-disposed when the calling component unmounts, because they are bound to the active scope (the component's scope, which Pinia's action helper sets as the current scope).

2. After an `await sleep(...)` in your `init()` action, execution has left Pinia's action wrapper and is now in a microtask continuation. The active scope is no longer the component's scope. In Vue 3.4, watchers created in this continuation were accidentally re-attached to the parent scope through the now-deprecated behavior described in vuejs/core#11496 (effect-scope inheritance across async boundaries).

3. Vue 3.5 deliberately removed that accidental inheritance. Reactive effects created after an `await` no longer bind to the parent component's scope. Instead they bind to the global / detached scope, which means they live until you explicitly stop them with `scope.stop()` or the watch handle returned by watch().

This is exactly what you observed: the watcher keeps firing because it is now leaked past unmount, AND it is created multiple times because each re-mount runs init() again and adds another watcher to the global scope (because the previous ones were never disposed).

The "move watch up before await" workaround you found works because then the watcher is created synchronously inside the action, which makes Pinia bind it to the component's effect scope correctly. After the await, that binding would not happen.

Two solutions, in order of preference:

Solution A (recommended): Capture the current scope before the await and explicitly pass it to watch():

```ts
import { defineStore } from 'pinia'
import { computed, watch, getCurrentScope, effectScope } from 'vue'

export const useDesignStore = defineStore('designEditorStore', () => {
  const designId = ref(0)
  const router = useRouter()

  async function init() {
    // Capture the component's effect scope BEFORE any await.
    // This is what Pinia was doing for you synchronously; we replicate
    // it explicitly so the post-await watcher binds correctly.
    const scope = getCurrentScope() ?? effectScope()

    const design = computed(() => router.currentRoute.value.params.designId as string)
    await sleep(1000)

    // Use scope.run() to ensure the watcher registers inside the captured scope.
    const stop = scope.run(() =>
      watch(design, (val) => {
        console.log('14 watch', val)
        designId.value = Number(val)
      }),
    )

    // Optional: also expose stop so the component can call it on unmount.
    return stop
  }

  return { designId, init }
})
```

And in your component:

```vue
<script setup>
const store = useDesignStore()
let stopWatcher: () => void

onMounted(async () => {
  stopWatcher = await store.init()
})

onBeforeUnmount(() => {
  stopWatcher?.()
})
</script>
```

Solution B (the canonical Pinia pattern, no scope plumbing): lift the watcher registration out of the async action. Setup-style stores support a clean pattern where you register watchers at store creation time, not inside actions:

```ts
import { defineStore } from 'pinia'
import { ref, computed, watch } from 'vue'

export const useDesignStore = defineStore('designEditorStore', () => {
  const designId = ref(0)
  const router = useRouter()
  const routeDesignId = computed(() =>
    router.currentRoute.value.params.designId as string,
  )

  // Watcher registered at store creation time. Because Pinia creates
  // the store inside the calling component's effect scope (when the
  // component calls useDesignStore() in setup()), this watcher is
  // auto-disposed when the component unmounts. No async, no problem.
  watch(routeDesignId, (val) => {
    designId.value = Number(val)
  })

  async function init() {
    await sleep(1000)
    // init() no longer creates the watcher; the watcher above was
    // registered when the store was first instantiated.
  }

  return { designId, init }
})
```

This is the cleaner pattern because it makes the data flow obvious: the store has a reactive dependency on the route, and you do not have to think about scope lifecycle at all. The async action just does its work; the watcher is set up once and disposed when the component unmounts.

Solution C (if you must keep the watcher inside the async action for some reason): subscribe to the watcher handle and dispose it explicitly when the action's promise resolves or the component unmounts:

```ts
async init() {
  const design = computed(() => router.currentRoute.value.params.designId as string)
  await sleep(1000)
  const handle = watch(design, (val) => {
    designId.value = Number(val)
  })
  // Store the handle somewhere the component can read it.
  this._watcherHandle = handle
}
```

Then in the component, call handle() in onBeforeUnmount. This is the no-helper version of Solution A.

The reason Solution B is best: it removes the entire problem class. As soon as you stop registering effects inside async code paths, you stop hitting this category of bug. Solution A is fine but you still have to remember to expose the stop handle to the component, which is easy to forget.

Why your reproduction fires "many times" specifically (not just once leaked): because each time HelloWorld mounts, it calls store.init() inside onMounted. The first mount creates a watcher after the sleep; the watcher runs and updates designId. When you navigate away, HelloWorld unmounts but the watcher leaks (Vue 3.5 bug for your pattern). When you navigate back, HelloWorld remounts, calls init() again, and a second watcher is created. Now you have two watchers, both running on every route change. Repeat and you get many.

In Vue 3.4, both watchers were auto-disposed when the component unmounted because the scope inheritance across await accidentally still worked. In 3.5, neither is disposed, so they accumulate. That is the bug, and that is what you have observed in the demo video.

Cross-references:
- Pinia issue #2786 (https://github.com/vuejs/pinia/issues/2786)
- Vue 3.5 changelog entries quoted above
- vuejs/core#11496 (the underlying effect-scope tightening)
- Pinia setup-style stores docs (https://pinia.vuejs.org/core-concepts/#setup-stores) — Solution B is the recommended pattern
